bgp: Add support for ClusterPool pod CIDRs

gandro opened the pull request:
This commit extends the BGP pod CIDR announcement logic to also parse
pod CIDRs used by ClusterPool IPAM. In contrast to the Kubernetes IPAM
mode (where pod CIDRs are stored native Kubernetes Node resources),
ClusterPool IPAM stores the allocated pod CIDRs in the CiliumNode
object.

Therefore, this commit extends the BGP speaker logic to also receive
CiliumNode updates.

gandro committed:
bgp: Add support for ClusterPool pod CIDRs

This commit extends the BGP pod CIDR announcement logic to also parse
pod CIDRs used by ClusterPool IPAM. In contrast to the Kubernetes IPAM
mode (where pod CIDRs are stored native Kubernetes Node resources),
ClusterPool IPAM stores the allocated pod CIDRs in the CiliumNode
object.

Therefore, this commit extends the BGP speaker logic to also receive
CiliumNode updates.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

ldelossa commented on daemon/cmd/daemon.go:
```diff
@@ -460,7 +460,12 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
+		case ipamOption.IPAMKubernetes:
+			d.k8sWatcher.NodeChain.Register(d.bgpSpeaker)
+		case ipamOption.IPAMClusterPool:
+			d.k8sWatcher.CiliumNodeChain.Register(d.bgpSpeaker)
```
Are these events invoked in the same fashion as the Node Events? Will be reliably get "NodeDelete" events for CiliumNode objects as well?

gandro replied:
Are there specific requirements around deletion? Generally speaking, we do assume that CiliumNode objects are deleted eventually, since otherwise the pod CIDRs assigned to them won't be available to any other nodes. There is an outstanding bug regarding this: https://github.com/cilium/cilium/issues/17837 - but in principle we rely on CiliumNode objects getting deleted for IPAM correctness too.

Overall, I'm not sure how timely these deletion events will occur. I assume what will happen in practice is that the pod CIDR routes to stale nodes are kept around until CRD garbage collection kicks in. However, since the pod CIDR (from which the routes are derived) will also not be released while the CRD is alive, there should be no danger of conflicting routes.

ldelossa replied:
It's a pretty soft requirement TBH. For full context:

The BGP code will see a NodeDelete object and attempt a graceful withdrawal of BGP routes from its peer. 

However, if the node goes off line, the peer will understand that and perform its own withdrawal of the routes. 

So, given your comments, I think we are OK either way. Its just that for graceful withdrawal we'd want to see CiliumNode delete events when the node is being taken out of the cluster.

christarazi replied:
FWIW, CiliumNode deletion relies on the ownerReference to be set in the CiliumNode resource pointing to the corresponding Node. In other words, when an ownerReference is set, the deletion of the referred will also delete the referee.

ldelossa reviewed (changes requested):
Looking pretty good. Are any tests necessary ?

gandro committed:
bgp: Add unit test for CiliumNode events

This adds a unit test which ensures correct handling of pod CIDR
announcements via CiliumNode events.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

ldelossa reviewed (approved):
LGTM

christarazi commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -374,6 +375,103 @@ func (s *MetalLBSpeaker) OnDeleteNode(node *v1.Node) error {
+	if node.GetName() != nodetypes.GetName() {
+		return nil // We don't care for other nodes.
+	}
+	var (
```
Since this code is essentially a copy of the K8s Node version, we should probably de-dup this by creating a new function which is agnostic to the type of the input object (CN or K8s Node) and just works on raw fields like meta & CIDRs. We still need to keep the CiliumNode handlers, but they should just call down to a function like below:

```
// feel free to use better names than I did; this is half real code, half pseudo.
func (s *MetalLBSpeaker) reconcileAdd(meta types.ObjectMeta, cidr CIDRType) error {
	if s.shutDown() {
		return ErrShutDown
	}
	if meta.GetName() != nodetypes.GetName() {
		return nil // We don't care for other nodes.
	}
	...
```

gandro replied:
Good feedback, thanks! I thought about this as well, but didn't realize everything besides the CIDRs relied on the `ObjectMeta`. Addressed, PTAL.

christarazi reviewed (changes requested):
Thanks for the PR. One comment about the duplication of code and how we can address it.

gandro commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -277,65 +279,46 @@ func (s *MetalLBSpeaker) OnUpdateEndpointSliceV1Beta1(eps *slim_discover_v1beta1
 		return ErrShutDown
 	}
-	if node.GetName() != nodetypes.GetName() {
+	if nodeMeta == nil || nodeMeta.GetName() != nodetypes.GetName() {
```
reviewer note: This nil check is added to safeguard against panics. I don't expect this to be hit in practice, but the unit tests will pass `nil` nodes to the node event handlers.

ldelossa commented on pkg/bgp/speaker/events.go:
```diff
@@ -26,6 +27,21 @@ const (
 	Delete
 )
 
+func (o Op) String() string {
```
:+1:

ldelossa commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -277,65 +279,46 @@ func (s *MetalLBSpeaker) OnUpdateEndpointSliceV1Beta1(eps *slim_discover_v1beta1
-// OnAddNode notifies the Speaker of a new node.
-func (s *MetalLBSpeaker) OnAddNode(node *v1.Node) error {
+// reconcileNode notifies the speaker of a node (K8s Node or CiliumNode) event
+func (s *MetalLBSpeaker) reconcileNode(op Op, nodeMeta *metav1.ObjectMeta, podCIDRs *[]string, withDraw bool) error {
```
nit: 

keep function name consistent with comments. 

```
func (s *MetalLBSpeaker) notifyNodeEvent(op Op, nodeMeta *metav1.ObjectMeta, podCIDRs *[]string, withDraw bool) 
```

gandro replied:
I like that suggestion. Will change it accordingly!

ldelossa commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -277,65 +279,46 @@ func (s *MetalLBSpeaker) OnUpdateEndpointSliceV1Beta1(eps *slim_discover_v1beta1
-			"component": "MetalLBSpeaker.OnAddNode",
-			"node":      node.Name,
+			"component": "MetalLBSpeaker.reconcileNode",
+			"op":        op.String(),
```
:+1:

ldelossa reviewed (approved):
refactor to "reconcileNode" looks good (nice suggest @christarazi) . Have small nit about the name of the function - but it subjective so do what you'd like with it.

christarazi reviewed (approved):
:rocket:
